### PR TITLE
Added `JointRV`  and `MarginalDistribution` in stats

### LINKED ISSUE #14852: `Lambda` does not accept indexed symbols.
`Lambda` raises an TypeError when the variable argument passed is an indexed symbol. Example:
```
>>> x1 = Indexed('x', 1)
>>> x1.is_Symbol
False
>>> Lambda(x1, x1**2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/akash/sympy/sympy/core/function.py", line 1659, in __new__
    raise TypeError('variable is not a symbol: %s' % i)
TypeError: variable is not a symbol: x[1]
```

### PR DESCRIPTION
**Pull request by akash9712**
#### References to other Issues or PRs

Fixes #14852 Closes #14931 

#### Brief description of what is fixed or changed
Added `JointRV`, which allows the user to create a joint random variable
with each of its component varying over the real line. Also added
`JointDistributionHandmade` for this purpose. 
Added `MarginalDistribution` to allow random variables with components marginalized out. This also adds the ability to handle compound random variables by marginalizing over latent distributions, with or without a given condition.

* stats
    * Add `JointRV` to allow user to create Joint random variables
    * Add `MarginalDistribution`
    * Ability to handle compound distributions 

#### Other comments
Because the PR uses indexed symbols in `Lambda`, which raised an error, suitable modifications were made in core/function.py.

### CODE REVIEW

**Comment by akash9712** [comment]
Ping @Upabjojr @jksuom 
Can you please restart the travis build? The failures seem random.

**Comment by jksuom** [comment]
The failures seem related to JointRV. For example,
```
______________________ sympy.stats.joint_rv_types.JointRV ______________________
File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/sympy-1.2.1.dev0-py2.7.egg/sympy/stats/joint_rv_types.py", line 48, in sympy.stats.joint_rv_types.JointRV
Failed example:
    density(N1)(1, 2)
Expected:
    exp(-2)/(2*pi)
Got:
    exp(-1)/(2*pi)
```
and
```
_______________ sympy/stats/tests/test_joint_rv.py:test_JointRV ________________
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.4.6/lib/python3.4/site-packages/sympy-1.2.1.dev0-py3.4.egg/sympy/utilities/runtests.py", line 1264, in _timeout
    function()
  File "/home/travis/virtualenv/python3.4.6/lib/python3.4/site-packages/sympy-1.2.1.dev0-py3.4.egg/sympy/stats/tests/test_joint_rv.py", line 71, in test_JointRV
    str(sqrt(2)*exp(-S(1)/2)*erf(sqrt(2)*x1/2)/(4*sqrt(pi)))
AssertionError
```

File: sympy/stats/joint_rv.py [hunk 1]
```diff
@@ -30,7 +32,7 @@ def __new__(cls, sym, dist):
 
     @property
     def set(self):
-        return self.domain.set
+        return self.distribution.set
```
Comment (Upabjojr) [inline sympy/stats/joint_rv.py]: why this?

Comment (akash9712) [reply sympy/stats/joint_rv.py]: I think self.domain should return an object of `RandomDomain`, which is not yet implemented for `JointPSpace`.

**Pull request closed without merging**